### PR TITLE
[jtag_riscv_agent] Added driver and register adapter

### PR DESCRIPTION
**Pull request by NigelScales**
- Added jtag_riscv_driver
- Added jtag_riscv_register_adapter
- Implemented jtag_riscv_monitor
- Fixed jtag_monitor (DR and IR were << 1)

Signed-off-by: Nigel Scales <nigel.scales@gmail.com>

### CODE REVIEW

**Comment by NigelScales** [comment]
Builds with VCS and Xcelium.
Not tested yet.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:33]: looks like there is not overlapping between 2 items. we don't need to save items in a queue.
```suggestion
      drive_jtag(rsp);
      seq_item_port.item_done(rsp);
```

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:20]: nit: one empty line is enough

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:110]: These are more like output
```suggestion
      output bit [DMI_OPW-1:0] status, output bit [DMI_DATAW-1:0] rdata);

```

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:130]: looks like most of functionality is ported from the seq. can you also update the seq in this PR?

**Review by weicaiyang — COMMENTED** [review]
Thanks for implementing this adapter approach. please continue to test it with vcs and xcelium.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (NigelScales) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:33]: I'll let you know how it works.
I think queuing is probably the safer option and will ensure back to back traffic.

Comment (NigelScales) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:35]: I think it's fine as is

**Comment by NigelScales** [comment]
I've tested it with  lc_ctrl.
I've cross merged this branch into lc_ctrl_jtag_csr and after demoting the register model warnings about field accesses csr_rw passes using JTAG.
To try add --run-opts +jtag_csr=1 to the command.
Currently this will replace the register adapter for default_map.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_reg_adapter.sv
Comment (cindychip) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_reg_adapter.sv:31]: Maybe later when can consider changing this from UVM_LOW to UVM_MEDIUM/HIGH?

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:35]: sorry, I'm not sure we need a queue to ensure back2back traffic. There is no overlap between 2 items.

With the queue, it also works but we need to a separate thread `drive_jtag` to run a forever loop. Without the queue, the code may be more clear.

**Comment by NigelScales** [comment]
@weicaiyang 
I've redone it without the queue because I want to be able to terminate the process easily if we get a reset.
You're right it wasn't needed.

File: hw/dv/sv/jtag_agent/jtag_monitor.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_agent/jtag_monitor.sv:38]: Use `%0h` and `%0d` to avoid ugly empty spaces.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_agent.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_agent.sv:34]: I made this comment in your other PR that follows this one, but really the change has to happen in this PR. `cfg` is already a handle of the JTAG sequencer. `cfg` has a field `in_reset`, which is set by the monitor when a reset event is detected. You could spawn a task in the sequencer's run_phase that stops the sequences whenever `cfg.reset` is set. IMO we should do this in the base sequencer `dv_base_sequencer` so that ALL agents benefit from it.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_agent_cfg.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_agent_cfg.sv:15]: I think this is already set to 1 by default.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:20]: As mentioned in the earlier comment, this code needs to be moved to the `sequencer`, where it looks at `cfg.in_reset` to stop all sequences.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:18]: This task is not called on every reset. See https://github.com/lowRISC/opentitan/blob/e10dfaf312614ae9ebcdfb8c79840de4cf45be7e/hw/dv/sv/dv_lib/dv_base_driver.sv#L21. 

IMO this process handle will likely always be null. 

Alternately, you could add a third fork in `get_and_drive()` that monitors the reset and then disable the `disable` the `drive_jtag_process` thread.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:38]: As suggested in comment above:
```systemverilog
          begin: monitor_reset
            @(cfg.in_reset);
            disable drive_jtag_process;
          end
```

`cfg.in_reset` is set by the monitor which monitors a reset event.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:77]: nit: strange formatting

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:124]: nit: Replace with `do..while`?
```systemverilog
  do begin
      bit [DMI_DRW-1:0] dout;
      send_csr_dr_req(DmiStatus, 0, 0, dout);
      status = dout[DMI_OPW-1:0];
      rdata = dout[DMI_DATAW-1:DMI_OPW];

      // The DmiInProgress status is sticky and has to be cleared by dmireset via DTMCS.
      if (status == DmiInProgress) begin
        send_riscv_ir_req(JtagDtmCsr);
        send_dtmcs_dr_req(DmiReset);
        send_riscv_ir_req(JtagDmiAccess);
      end
  end while (status == DmiInProgress);
```
Does this mean the bottom 3 bits of `rdata` which is `DMI_DATAW` wide will always be 0?

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:110]: Could you use named arguments so that it is clear what args are we setting?

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_monitor.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_monitor.sv:23]: Please run this file through the `verible-verilog-format` formatter so that the style lint issues are resolved.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_reg_adapter.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_reg_adapter.sv:29]: Indentation is off. Recommend formatting this way:
```systemverilog
    `DV_CHECK_RANDOMIZE_WITH_FATAL(
        jtag_item,
        if (local::rw.kind == UVM_WRITE) {
          op == DmiWrite;
        } else {
          op == DmiRead;
        }
        // Convert to word address by slicing
        addr == local::rw.addr[DMI_ADDRW + DMI_WORD_SHIFT - 1 : DMI_WORD_SHIFT];
        data == local::rw.data[DMI_DATAW-1:0];
    )
```

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_reg_adapter.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_reg_adapter.sv:35]: nit: empty newline not needed

**Review by sriyerg — CHANGES_REQUESTED** [review]

**Comment by sriyerg** [comment]
I left quite a few comments. Please amend existing commits directly rather than creating new commits. You can follow these steps:
```console
git checkout jtag_riscv_driver # If not already in this branch
git rebase -i master

# This will open a text editor with the 3 commits
# Replace 'pick' with 'edit' to edit that particular commit. 
# git-rebase will rebase each commit one-by-one starting
# from the top and perform the action. If commit is marked 
# as edit, then it will stop the rebase to let you amend the 
# the commit. Address the review comments for that particular
# commit. Once done:
git add -u # Add the updated files. DO NOT EDIT anything other than review comments.
git commit --amend # Amend this commit. 
git rebase --continue

# git-rebase will continue and stop at the next commit marked 
# for editing. Address the comments for this commit and 
# follow the steps above to move on to the next. 
# 
# Continue these steps until ALL commits are rebase and 
# you see the message "Rebased successfully". 
```

**Comment by sriyerg** [comment]
> I left quite a few comments. Please amend existing commits directly rather than creating new commits. You can follow these steps:
> 
> ```
> git checkout jtag_riscv_driver # If not already in this branch
> git rebase -i master
> 
> # This will open a text editor with the 3 commits
> # Replace 'pick' with 'edit' to edit that particular commit. 
> # git-rebase will rebase each commit one-by-one starting
> # from the top and perform the action. If commit is marked 
> # as edit, then it will stop the rebase to let you amend the 
> # the commit. Address the review comments for that particular
> # commit. Once done:
> git add -u # Add the updated files. DO NOT EDIT anything other than review comments.
> git commit --amend # Amend this commit. 
> git rebase --continue
> 
> # git-rebase will continue and stop at the next commit marked 
> # for editing. Address the comments for this commit and 
> # follow the steps above to move on to the next. 
> # 
> # Continue these steps until ALL commits are rebase and 
> # you see the message "Rebased successfully". 
> ```

To make this simpler - please `squash` the 3 commit into a single commit. All these updates belong to a single commit, as the PR title suggests.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (NigelScales) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:77]: That is actually the result of running verible-format.py
It makes a mess of macros.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_agent.sv
Comment (NigelScales) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_agent.sv:34]: I'm now monitoring in_reset and using it to terminate activity in the driver.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (NigelScales) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv]: "Does this mean the bottom 3 bits of rdata which is DMI_DATAW wide will always be 0?"
Not quite but it is a bug.
Top 2 bits of read data will be 0,
Not detected due to lack of registers to read (exclusions).

image

**Comment by NigelScales** [comment]
@sriyerg 
What I've done before is squash the commits on the final PR merge.
All scripted by GitHub.
Expecting git newbies to do rebasing is a bit of a stretch.
The chances of it going wrong are very high.

**Comment by rswarbrick** [comment]
@NigelScales: You're right that it would be unwise to trust beginning contributors not to mess up git rebasing. Which is why we don't merge things before we've taken a look at the changes. We *do* allow PRs with multiple commits: this is sometimes exactly what you want to appear in the eventual log. This means we wouldn't want an "auto-squash" feature like you're describing.

**Comment by NigelScales** [comment]
@rswarbrick 
I'm not too bad at git but I've never used rebase.
I've managed to get by with merge.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:9]: we probably don't need reset_occured since you use cfg.in_reset

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:69]: this `cfg.in_reset` is added everywhere, which doesn't look good.

I would prefer to let the low-level driver (jtag_driver) to handle the reset. (jtag_driver has yet to add support for reset). In other low-level driver, we finish all the items immediately, once reset occurs. In upper-level, we could just continue sending the item even it's in reset, as all the items will be finished immediately. 
The only place we need to add `cfg.in_reset` is the while loop.

The other approach is to use
```
// kill drive_jtag once reset occurs
`DV_SPINWAIT_EXIT(drive_jtag,
           wait(cfg.in_reset))
```

WDYT? @sriyerg

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [inline hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:143]: why not use output, this task doesn't use the input value of status and rdata.
explicitly using output is better than passing the handles 
```suggestion
  protected virtual task check_csr_req_status(output bit [DMI_OPW-1:0] status,
                                              output bit [DMI_DATAW-1:0] rdata);
```

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_agent.sv
Comment (sriyerg) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_agent.sv:34]: Ohh, you need the handle to the vanilla JTAG sequencer within the riscv driver. This is fine I suppose.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (sriyerg) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:69]: Yes, what you suggest (IIUC) will make the code significantly simpler (i.e. kill the `drive_jtag()` thread externally if reset is encountered).

File: hw/dv/sv/jtag_riscv_agent/seq_lib/jtag_riscv_csr_seq.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/seq_lib/jtag_riscv_csr_seq.sv:11]: The `dv_utils_pkg::bus_op_e` provides a common typedef to represent read / write operations. Fine with keeping this for now and optimizing later though.

File: hw/dv/sv/jtag_riscv_agent/seq_lib/jtag_riscv_csr_seq.sv
Comment (sriyerg) [inline hw/dv/sv/jtag_riscv_agent/seq_lib/jtag_riscv_csr_seq.sv:30]: `exp_err` is a bit which is already initialized to 0. Can stick to `uvm_object_new`. But, fine to do it later though.

**Review by sriyerg — APPROVED** [review]
Looks great, thanks! Left 2 more comments - fine with addressing them later. Will defer to Weicai for +1 approval and merge.

**Comment by NigelScales** [comment]
@sriyerg it now appears this branch is now unmergeable.
I think this is due to re-writing history.
From now on I'm going to avoid that and rely on a final squash merge when you merge the PR.

**Comment by NigelScales** [comment]
> @sriyerg it appears this branch is now unmergeable. I think this is due to re-writing history. From now on I'm going to avoid that and rely on the final squash merge when you merge the PR.

I've done a 
git reset --hard <old commit>
git push -f

To get back to a mergeable state.

**Comment by NigelScales** [comment]
@sriyerg 
I can't merge this I don't have write permissions.

**Comment by weicaiyang** [comment]
> @sriyerg I can't merge this I don't have write permissions.

@NigelScales yes, only the people in this [list](https://github.com/lowRISC/opentitan/blob/master/COMMITTERS) can approval and merge the PR.

there are still a few unaddressed comments. Can you click the `file changed` tab and take a look? Looks like those comments can be addressed very quickly. Once those are done, I can merge the PR.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (NigelScales) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:69]: Each task spawns a new sequence so the code is code is correct.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (NigelScales) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:143]: Fine as is.

**Comment by NigelScales** [comment]
Hi Guys,
I'm happy with the code as is - it works and I have no more time to spare on it..
Can we get this merged please it's needed for the next PR.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:69]: @NigelScales why not try to make the reset part simpler. This is the first high-level agent in the repo. It's nice to create a good example for others to follow. Some other blocks will also reuse this agent, so making it more clean and readable will be helpful to others.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:9]: oh, looks like this is needed to drive DmiHardReset.

File: hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv
Comment (weicaiyang) [reply hw/dv/sv/jtag_riscv_agent/jtag_riscv_driver.sv:143]: `ref` works, but `output` is probably slightly better.

**Review by weicaiyang — APPROVED** [review]
Overall LGTM, will create an PR to address the last 2 comments

**Pull request merged**
